### DISCUSSION TITLE
How to cache the Strapi GraphQL queries?

### DISCUSSION THREAD

**Discussion by abdonrd** [strapi/strapi#7499] [Legacy Topics]
Since the information in my DB doesn't change very often, I wanted to cache the queries.

How to cache the GraphQL queries? With something like Redis or so.

**Comment by derrickmehaffy** [comment]
I don't believe there is a solution to this currently (at least not within Strapi by default) for the REST side there is a community middleware: https://www.npmjs.com/package/strapi-middleware-cache but this won't work for GraphQL, most likely this would need to be custom built and integrated into the apollo-server.

**Comment by abdonrd** [comment]
There are any option to use this? Extending the schema:

https://www.apollographql.com/docs/apollo-server/performance/caching/#adding-cache-hints-statically-in-your-schema

**Reply by abdonrd** [reply]
This doesn't work:

`api/matchday/config/schema.graphql.js`

```javascript
module.exports = {
  definition: `
    extend type Matchday @cacheControl(maxAge: 240) {}
  `,
};
```

**Comment by abdonrd** [comment]
And related with saving full responses to a cache:

https://www.apollographql.com/docs/apollo-server/performance/caching/#saving-full-responses-to-a-cache

PR related: https://github.com/strapi/strapi/pull/7519

**Comment by iicdii** [comment]
Basically, the issue was with `makeExecutableSchema` from `apollo-server-koa`. `makeExecutableSchema` is not compatible with apollo server cache.

https://github.com/strapi/strapi/blob/435b294439b2ddb54e79d1f825a7de4f97482116/packages/strapi-plugin-graphql/services/schema-generator.js#L9
https://github.com/strapi/strapi/blob/435b294439b2ddb54e79d1f825a7de4f97482116/packages/strapi-plugin-graphql/services/schema-generator.js#L98-L101

You can find reference in https://www.apollographql.com/docs/apollo-server/performance/caching/#defining-cache-hints

> Important note on compatibility: **Setting cache hints is currently incompatible with the graphql-tools implementation** of schema stitching, because cache hints are not appropriately communicated from one service to the other.

The solution is defining custom directive in our schema. You can use cache-control directive by defining them in GraphQL extension.
Create custom `schema.graphql.js` in `extensions/graphql/config/schema.graphql.js`
```graphql
enum CacheControlScope {
  PUBLIC
  PRIVATE
}

directive @cacheControl (
  maxAge: Int
  scope: CacheControlScope
) on FIELD_DEFINITION | OBJECT | INTERFACE
```

Then you can use directive in definition something like below
```graphql
type Author @cacheControl(maxAge: 60) {
  id: Int
  firstName: String
  lastName: String
  posts: [Post]
}
```
